model_compression_toolkit/core/common/mixed_precision/solution_refinement_procedure.py

```diff
@@ -108,40 +109,106 @@ def greedy_solution_refinement_procedure(mp_solution: List[int],
+        valid_candidate_for_activation = candidate_activation_bits > current_activation_bits
+        return valid_candidate_for_weight and valid_candidate_for_activation
+
+    valid_candidate_for_weight = all(candidate_weight_bits[attr] == current_weight_bits[attr] for attr in current_weight_bits.keys())
+    valid_candidate_for_activation = candidate_activation_bits > current_activation_bits
+    return valid_candidate_for_weight and valid_candidate_for_activation
+
+def is_weight_restriction_compliant(
+        current_activation_bits: int,
+        current_weight_bits: Dict[str, int],
+        candidate_activation_bits: int,
+        candidate_weight_bits: Dict[str, int],
+        is_activation_restricted: bool
+) -> bool:
+    """Used when we try to improve weights bit-width. The comparison of the activation bit width
+    is dependent on whether the target RU limits activation usage, thus we look for a MP solution
+    for activations, or not."""
+
+    if is_activation_restricted:
+        valid_candidate_for_weight = all(candidate_weight_bits[attr] > current_weight_bits[attr] for attr in current_weight_bits.keys())
+        valid_candidate_for_activation = candidate_activation_bits >= current_activation_bits
+        return valid_candidate_for_weight and valid_candidate_for_activation
+
+    valid_candidate_for_weight = all(candidate_weight_bits[attr] > current_weight_bits[attr] for attr in current_weight_bits.keys())
+    valid_candidate_for_activation = candidate_activation_bits == current_activation_bits
+
+    return valid_candidate_for_weight and valid_candidate_for_activation
+
+
+def get_candidate_bits(candidate) -> Tuple[Dict[str, int], int]:
     """
-    Find node's valid candidates to try and improve the node's MP chosen candidate.
-    Valid indices are indices of candidates that have higher number of bits for both weights and activations
-    (if they are quantized in this node).
+    Extract weight and activation bits from a candidate.
+
+    Returns:
+        Tuple of (weights_n_bits, activation_n_bits)
+    """
+
+    weight_attrs = candidate.weights_quantization_cfg.all_weight_attrs
+    weight_attrs_to_nbits = {}
+    for w_attr in weight_attrs:
+        weight_attrs_to_nbits[w_attr] = candidate.weights_quantization_cfg.get_attr_config(w_attr).weights_n_bits
+    activation_n_bits = candidate.activation_quantization_cfg.activation_n_bits
+    return weight_attrs_to_nbits, activation_n_bits
+
+
+def _get_valid_candidates_indices(
+        node_candidates: List['CandidateNodeQuantizationConfig'],
+        current_chosen_index: int,
+        is_activation_restricted: bool,
+        is_weight_restricted: bool
+) -> List[int]:
+    """
+    Find node's valid candidates to improve the node's MP chosen candidate.
+
+    Valid indices are indices of candidates that have higher number of bits for both
+    weights and activations (if they are quantized in this node).
```

Comment:
Find valid candidate indices for a node to enhance its MP-selected candidate. 
A candidate is considered valid if it has a higher bit-width for both weights 
and activations that are quantized in this node.